### ISSUE TITLE
Low cardinality on wp_actionscheduler_actions table

### ISSUE THREAD

**Issue by 84em** [woocommerce/action-scheduler#658]
There's an index on the status column with a cardinality of 3.  It's my understanding low cardinality indexes can affect mysql performance negatively.  Curious if this has been addressed before, and/or benchmarks run?

**Comment by barryhughes** [comment]
Hi @84em,

So I'll start by noting that when we migrated from storing scheduled actions in the post table to the current schema, we observed an improvement of around 70%.

However...if I'm understanding your question correctly, you're specifically asking if we've benchmarked performance in terms of defining the `actionsheduler_actions.status` column as an index (vs not as an index), is that correct? If so ... I'm not sure that that specifically _has_ been benchmarked, but I can certainly check in more detail.

Have you performed any benchmarking (or rough testing of your own) in relation to this that you're happy to share?

**Comment by 84em** [comment]
Hi Barry,

Thanks for the follow-up.  I haven't benched marked yet.

I reached out as I'm dealing with a very busy Woo Subscriptions site that relies heavily on Action Scheduler.  Lots of stuff runs through Action Scheduler from a variety of high load plugins.

We had been experiencing a lot of MySQL Deadlocks on the actions table.  Hundreds, sometimes thousands, every day.  They all looked like this:

`WordPress database error Deadlock found when trying to get lock; try restarting transaction for query UPDATE wp_actionscheduler_actions SET attempts = attempts+1, status='in-progress', last_attempt_gmt = '2021-02-24 14:39:37', last_attempt_local = '2021-02-24 08:39:37' WHERE action_id = 3372695 `

So I was kind of grasping at straws trying to find any performance wins and seemed to recall from my old pre-WordPress days that low cardinality indexes can affect database performance negatively.  

I never did try removing the index.  But what I did find was a PR specifically addressing deadlocks:

https://github.com/woocommerce/action-scheduler/pull/535/commits/904c37e937d00f531d77ffec244d3eedad8927a6

I applied that change to our actions table and the deadlocks went away immediately.

So TL;DR -- consider this my big +1 for getting that PR merged into Action Scheduler core.  :)

**Comment by barryhughes** [comment]
That is a useful note, thank you @84em :+1:

**Comment by barryhughes** [comment]
Noting #535 was merged in. I'm going to go ahead and close this issue, let me know if for any reason it should be re-opened.
